### ISSUE TITLE
Speeding up package rebuild

### ISSUE THREAD

**Issue by brupelo** [kayhayen/Nuitka#121]
Let's say I build meaty packages such as numpy, pyopengl... building these packages take ages, and well... that's fine, at the end of the day if you're doing a "cold build" there isn't nothing to complain about as nuitka has to do a lot of job to deliver.

BUT! if we rebuild the same packages again (without modifying not a single line of the python package) I would expect nuitka of course to parse again the whole python source (unless you cache compiler flags somehow) and then applying optimizations to get in shape to start writing c source files, isn't? 

I see the process is not optimal though, you're removing the existing c/obj files no matter what... why? I mean, yeah... you keep the algorithm easy but... why? :D

Alright, please consider the next "trivial" improvement:

 - Let's say you've ended up to parse the whole python package and you're ready to generate the c source code, you haven't deleted any c/obj files in the build folder.
 - So, before deleting any c/obj source file, you should compute the checksum of each c file before dumping to disk, if the checksum is different to the existing file disk only in that case you can overwrite the existing c file, otherwise, skip it. Why? Important thing here is preserving file timestamps so the objs won't need to be generated if they aren't need to.
 - Once you've overwritten only the c files you'll have a list with all c files, so using that list you'll be able to remove the rest of unused files (c/obj files)

Thanks to this improvement, we should be able to skip the whole compilation if timestamps haven't changed at all, so basically the compiler would have to do nothing and we could link the whole thing again for free.

Does it make sense?

**Comment by kayhayen** [comment]
I wrote about caching the mailing list once, it's a multi tiered thing, where the object files are only part of the equation. Storing module level optimized XML of modules is also a part, yes. On Linux I kind of expect you to use ccache. On Windows and MSVC I would integrate dlcache, but nobody took this on despite repeated requests of mine.

I myself will probably focus on "libnuitka" as I think that for deployment of multiple programs that kind of a must have, and it also will save re-compilation and even re-consideration for re-compilation for a fair bit of code. On Windows most so, because the one measure I do have in place won't work.

On Windows the build folder contains a scons level cache, where files are remembered with checksum and everything. That scons cache seems non-transferable or possible to share. The real solutions would be ccache for MinGW64 and dlcache for MSVC, but when I tried to get at binaries for "ccache" I failed, and I think it's not trivial to build, needs "make" which again is hard to come by.

To me building on Windows is a pain, as many of my commits do not change the created source or not for all test cases, and the number of test suites that I can run has hit a limit. On the other hand I should try and do within the project only things that I feel currently only I can do. I am trying to get more involvement via Twitter and so far it seems to happen somewhat.

**Comment by brupelo** [comment]
Couple of question:

 - What is this "libnuitka" you're talking about? The word itself sounds really interesting and like some sort of massive ammount of code shared across python projects that will speed up compilations...
 - Could you please point me out to some spots in the source code where you delete the c/obj files and you dump the c files? Eventually, I'd like to make some little experiments about it.

In any case, my whole point is, if a c file hasn't changed, don't overwrite it and don't delete the obj file, don't compile it again, for instance:

    a.cpp  a.obj  (updated)
    b.cpp  b.obj  (updated)
    c.cpp  c.obj   (outdated)
    d.cpp d.obj   (unused)

Let's say just c.cpp needs to be recompiled, you would just need to do:

    1) Delete c.cpp/c.obj, dump to disk c.cpp and include c.cpp in the SCons files pending to compile
    2) Link a.obj/b.obj/c.obj

Good thing about the way you're creating the source folder is inter-dependencies amongst c files and headers is minimal so you don't need fancy analizers such as [showIncludes](https://msdn.microsoft.com/en-us/library/hdkef6tk.aspx) to figure out recursively what needs to be updated... So I think this little tweak would allow to boost up compilation in many cases. Cos let me ask you, if I've got a python composed by {a.py, b.py, c.py}, all of them independent each other and I just tweak c.py, both a.py/b.py would generate the same source file, is this assumption correct or...?

Anyway, please let me know, i'd like to give it a shot to some ideas here :)

**Comment by kayhayen** [comment]
The libnuitka is an issue here too, #102 

The idea is that the compiled function type, etc. would only be compiled once, and be available for use, much like "libpython" to nuitka created programs and extension modules. That will save compilation time. Actually a lot more on Windows due to the lack of dlcache.

Having a compiled standard library or even Python distribution, would be a separate idea, that I would also welcome. I imagine working with something similar to AnaConda where the build process is then plugged to make the installation of "setup.py" go through "bist_nuitka". That is a very ambition thing, that I think will materialize once the C type work proves so valuable that people will want to have it effective as much as possible.

I want to make other changes before, that e.g. make Nuitka include the source in its output as well, and if changed, to override compiled code until the binary is updated.

As for your issue. On Linux ccache is working. On Windows, it would too, I just can't have it, and dlcache would work for MSVC, but nobody did the work yet despite repeated pleas in the past, but then it would work. For all I know, the location of the modules wouldn't matter with runtime "__file__" reference mode, and does not if you do not move sources. Virtualenvs in temp locations are killers for "original" "__file__" reference mode though.

Obviously I would gain a lot by having this dlcache integration, and users would too, although a lot less than I would. But I also got a compiler to write for it to make people care enough so they will eventually wind up doing that for me. And since developing on Linux is my normal thing, I am not going to do it for some time more myself.

**Comment by kayhayen** [comment]
As for your not deleting object files and re-using it. Yes, Nuitka deletes them, but it does NOT delete the Scons cache, or at least I am convinced of that. And Scons when it finds the source unchanged *will* take it out of the cache, much like ccache does.

So as long as you repeat commands, I would expect no C compilation to be invoked by scons. You can confirm that with --show-scons to be sure.

If that is not a true statement, it would be a Nuitka issue @brupelo which of course it may well be, since I didn't check all that recently.

**Comment by brupelo** [comment]
@kayhayen Wow! `--show-scons` was a total breakthrough to me here. That option explains quite a lot about what's going on and everything makes a little bit more sense now. Thanks to suggest to use it!

So yeah, after knowing that the cache I was referring to (obj file caching) was working already ok already I feel pretty dumb I have opened this issue in the first place :/... Next time before opening any issue with ideas/suggestions I'll make sure I totally understand first how nuitka works so my suggestions will make sense and I won't make a fool of myself, my bad, sorry about it.

Also, grepping the `--show-scons` not only allows to inspect in a really nice way what's cached or not but also allow you to see the final obj files in a very tidy way, so as a casual user I realize (even if the code generator subsystem is still a black box to me) what the big picture is. For instance, now it makes sense how `Nuitka\nuitka\build\static_src` fits into the big picture and I guess those 505kb of static source code will be included somehow with the the future `libnuitka` (sort of... I'm speculating here :)).

I've got some questions though, consider this little dummy library:

a.py

    def foo():
        print('foo')

b.py

    def bar():
        print('bar')

c.py

    from a import foo
    from b import bar

If I generate an extension module of `c.py` without using --recurse-all I'll get a 228kb c.pyd whose .obj files are:

    Retrieved `c.build\module.c.obj' from cache
    Retrieved `c.build\__constants.obj' from cache
    Retrieved `c.build\__constants_data.obj' from cache
    Retrieved `c.build\__frozen.obj' from cache
    Retrieved `c.build\__helpers.obj' from cache
    Retrieved `c.build\CompiledCellType.obj' from cache
    Retrieved `c.build\CompiledFunctionType.obj' from cache
    Retrieved `c.build\CompiledMethodType.obj' from cache
    Retrieved `c.build\CompiledGeneratorType.obj' from cache
    Retrieved `c.build\CompiledCoroutineType.obj' from cache
    Retrieved `c.build\CompiledAsyncgenType.obj' from cache
    Retrieved `c.build\CompiledFrameType.obj' from cache
    Retrieved `c.build\CompiledCodeHelpers.obj' from cache
    Retrieved `c.build\InspectPatcher.obj' from cache
    Retrieved `c.build\MetaPathBasedLoader.obj' from cache
    Retrieved `c.pyd' from cache

while If I use `--recurse-all` I'll get a 240kb c.pyd with this content:

    Retrieved `c.build\module.a.obj' from cache                        
    Retrieved `c.build\module.b.obj' from cache                        
    Retrieved `c.build\module.c.obj' from cache                        
    Retrieved `c.build\__constants.obj' from cache                     
    Retrieved `c.build\__constants_data.obj' from cache                
    Retrieved `c.build\__frozen.obj' from cache                        
    Retrieved `c.build\__helpers.obj' from cache                       
    Retrieved `c.build\CompiledCellType.obj' from cache                
    Retrieved `c.build\CompiledFunctionType.obj' from cache            
    Retrieved `c.build\CompiledMethodType.obj' from cache              
    Retrieved `c.build\CompiledGeneratorType.obj' from cache           
    Retrieved `c.build\CompiledCoroutineType.obj' from cache           
    Retrieved `c.build\CompiledAsyncgenType.obj' from cache            
    Retrieved `c.build\CompiledFrameType.obj' from cache               
    Retrieved `c.build\CompiledCodeHelpers.obj' from cache             
    Retrieved `c.build\InspectPatcher.obj' from cache                  
    Retrieved `c.build\MetaPathBasedLoader.obj' from cache             
    Retrieved `c.pyd' from cache                                       

We can see in the 2nd case both `module.a.obj` and `module.b.obj` have been included. Thing is, no matter if I use --recurse-all or not, in both cases the library is exposing correctly `foo` and `bar` functions with the right code... 

Questions:

 - What's the real difference of using `--recurse-all` or not? 
 - On which cases should I use --recurse-all?  
 - Is --recurse-all adding a lot of code duplication? See above, 1st case c.pyd is 228kb and 2nd is 240kb, both of them working ok... so, what's going on?

Conclusion: I guess this issue could be already tagged like "Question" and be closed? Not sure I've brought anything new to the table here and it ended up to be just me asking stuff... Hopefully you will be able to use this material in blog posts, FAQs, tutorials, videos or similar so more devs will benefit from it... Anyway, your call.

**Comment by kayhayen** [comment]
I am OK with you opening issues without perfect research upfront @brupelo so don't worry.

The recursion options control if you want to include a module that is imported in code statically or not dynamic enough at least, or not. For programs to be as much accelerated as possible, --recurse-all is best use. For extension modules, you probably sometimes want to include a dependency into the module, or if you use two name or more spaces like e.g. pywin32 does.

I am looking at rename the --recurse optons to auto-follow or something, but I didn't yet find anything that conveys their use all that much. They kind of are like --include-if-used in case of --recurse-to and hard to grasp for new users with current naming.

The --include options are newer and forcing and are most useful for dynamic code.

**Comment by brupelo** [comment]
@kayhayen Mmm, I think I understand now, thanks to clarify. In any case, I was asking because at least with personal projects where I've got perfect control about how I do all importing management I'll try to skip `--recurse-all` as much as possible. For instance, consider a pyqt program where I've got few hundreds of plugins. The idea is first to compile the exe only once (without --recurse-all) and then building my plugins independently.

One last question, I've already build many python extension modules and I've got them stored in a "3rdparty" folder, I've already checked that in case I want to compile a python script living in that folder that use those extension modules nuitka will check the extension modules are the "3rdparty" one and not the site-packages ones... So for instance, let's say I've built numpy in 1200s, I've already checked that if a program uses directly numpy.pyd, the compilation will take very few seconds, which makes total sense.

Anyway, my question is, what'd be the correct way so nuitka will look at the `3rdparty` folder before anything else? I've tried to add "3rdparty" into a `.pth` file in my virtualenv but that won't do it... Another choice would be tweaking PYTHONPATH although... do you recommend any better choice :/? Asking cos I really dislike having/tweaking PYTHONPATH at all and I always discourage python newcomers to use it in the first place. Although I've run out of ideas what's the best way to go about it... Thing is, I don't want to copy/move my python stuff into the 3rdparty folder, so... I'm all ears :)

**Comment by kayhayen** [comment]
@brupelo Can I ask you to put questions for new topics into new issues? That will make it easier to use the result afterwards. :-)

**Comment by blakdronzer** [comment]
I know i am quite new in here - this is the first time i am using nuitka. It is a great tool to have but when working on bigger projects - the compilation becomes a nightmare. On windows - 1 recompile if required - for me it takes 3-4 hrs of time. (No offenses - i know it aint your fault, the gcc compilation is the 1 that takes hell lot of time) That is a bit too much to wait. I understand it goes through a lot of conversions from .py to .cpp and then to obj / dll .. etc etc.. 

It might be possible you may have already looked in  but still i was wondering on 1 thing. When i go through the compilation process - looking at --show-scons, what i realized that it over and over kept building the obj files. Though what was noticed, all the cpp files were intact. I was wondering as why was it rebuilding the obj files for standard libraries again and again. Why can't those be left intact.

There are codes in the project that are used for imports - understood. This are technically the codes expected by the user to change and recompiling them makes sense. But the codes used as standard libraries that are found from site_packages folder are libraries that barely user goes and changes. 

I was wondering if this libraries (code / obj files and other that may be built along) be left intact in the build folder so - very 1 thing .. the rebuilding stuff could speed up by this? Is there such possibilities?

I do see the obj / cpp files being there even after the build was made. But unfortunately - still next when i start recompiling second time - it still compiles the very same objects saying it dose not exists.

**Comment by kayhayen** [comment]
@blakdronzer The missing caching issue is indeed annoying and it can be done, but nobody is working on it, but I would clearly welcome it. I think I broke down necessary steps before. Using `ccache` and `clcache` ought to help with unchanged code, the User Manual describes how to use those I think.

**Comment by deajan** [comment]
@kayhayen I'm currently searching for ways to optimize my Nuitka compilation times, see https://github.com/Nuitka/NUITKA-Utilities/issues/9#issuecomment-571586198 where I could divide time by 2.

I've noticed that the C files Nuitka generates are regenerated on every run, so I thought of a caching mechanism I could create, which would store the sha256sum of all source files that generate c files, and reuses the generated c file if the source has the same sha256sum, with let's say a LRU eviction mechanism or so.

I've then checked two different c files that come from the same source file compiled for two different programs, and noticed they differ because of some constants, eg `UNSTREAM_STRING_ASCII(&constant_bin[ XXXXXX ], YY, Z);`, where XXXXXX varies for same sources:


```
/* Function to create module private constants. */
static void createModuleConstants(void) {
    const_str_digest_33c72c5b0e9ba79b972f105c5aaaa097 = UNSTREAM_STRING_ASCII(&constant_bin[ 169631 ], 12, 0);
    const_tuple_3ce17fc0476de00dbb83c4b65c00080e_tuple = PyTuple_New(2);
    const_str_plain_NUITKA_PACKAGE_api = UNSTREAM_STRING_ASCII(&constant_bin[ 169643 ], 18, 1);
    PyTuple_SET_ITEM(const_tuple_3ce17fc0476de00dbb83c4b65c00080e_tuple, 0, const_str_plain_NUITKA_PACKAGE_api); Py_INCREF(const_str_plain_NUITKA_PACKAGE_api);
    PyTuple_SET_ITEM(const_tuple_3ce17fc0476de00dbb83c4b65c00080e_tuple, 1, const_str_digest_5bfaf90dbd407b4fc29090c8f6415242); Py_INCREF(const_str_digest_5bfaf90dbd407b4fc29090c8f6415242);
    const_str_digest_6eb7a56cbca1279a1c81e4b3885ca177 = UNSTREAM_STRING_ASCII(&constant_bin[ 169661 ], 15, 0);

    constants_created = true;
}
```

vs

```
/* Function to create module private constants. */
static void createModuleConstants(void) {
    const_str_digest_33c72c5b0e9ba79b972f105c5aaaa097 = UNSTREAM_STRING_ASCII(&constant_bin[ 178450 ], 12, 0);
    const_tuple_3ce17fc0476de00dbb83c4b65c00080e_tuple = PyTuple_New(2);
    const_str_plain_NUITKA_PACKAGE_api = UNSTREAM_STRING_ASCII(&constant_bin[ 178462 ], 18, 1);
    PyTuple_SET_ITEM(const_tuple_3ce17fc0476de00dbb83c4b65c00080e_tuple, 0, const_str_plain_NUITKA_PACKAGE_api); Py_INCREF(const_str_plain_NUITKA_PACKAGE_api);
    PyTuple_SET_ITEM(const_tuple_3ce17fc0476de00dbb83c4b65c00080e_tuple, 1, const_str_digest_5bfaf90dbd407b4fc29090c8f6415242); Py_INCREF(const_str_digest_5bfaf90dbd407b4fc29090c8f6415242);
    const_str_digest_6eb7a56cbca1279a1c81e4b3885ca177 = UNSTREAM_STRING_ASCII(&constant_bin[ 178480 ], 15, 0);

    constants_created = true;
}
```

Is there any way two different program compilation runs of the same source might endup with the same constant ? This would help create a global C cache instead of a per program C cache.

If so, I'd be glad you to point me to the right direction in the nuitka sources to write a c code cache, which would obviously speed up Nuitka by orders of magnitude.

Also, as a side question, I noticed CFLAGS have '-O3' set in SingleExe.scons
Is there a way to override -O3 and set -Os/-O0 or anything else perhaps ?

**Comment by kayhayen** [comment]
Just to chime in, the caching of Nuitka source code is very spoiled because of the constants. They are using a global blob, and indexes in there. That is very unreliable to remain constant even with small changes in other parts, e.g. to include a module more or less, it's not going to remain re-usable.

I have plans for a marshal like replacement. The bad thing about marshal is that it doesn't do all types, and therefor was not usable. Also a single global stream is vulnerable. I want to introduce it, because it will not generate C code, but use a sort of bytecode approach like marshal does. Also it would be the case, that the constants resolution has a single pointer, which gets initialized during program load, and constants would be accessible with module local indexes from it.

An extra benefit that I see is that the initial constants stuff, might not have to necessarily cache the resulted object (release it again, saves memory at run time, costs at run time to create on the fly, but for single use ones, that is much better), and create all of them at the start.

With that change, and only then, I would expect same module node tree -> same C source. Until then I fear all caching attempts are futile. But recent caches at least allow for `ccache` to do its job. I suspect that `clcache` may also be more successful.

For options, `CCFLAGS` is inserted, however, the `-O3` is appended, and may override. I guess, we would have to move it to the end in `SingleScons.exe` for you to be allowed to override it. Because `-Os -O3` is not going to work, but `-O3 -Os` is.

In my experience however, there is not much result coming from these gcc options.

Yours,
Kay
